### PR TITLE
Index with length equal to column length keeps generating diff

### PR DESCRIPTION
**Pull request by janedbal**
|      Q       |   A
|------------- | -----------
| Type         | bug
| Fixed issues | 0

#### Summary

If any of indexed columns has length equal to the varchar limit, diff always generates index re-creation SQLs (drop + create). 

-----------

This is easily fixable by passing null for those columns where I want to keep full length.

### CODE REVIEW

File: tests/Functional/Schema/MySQLSchemaManagerTest.php [hunk 1]
```diff
@@ -541,4 +541,27 @@ public function testEnsureTableWithoutOptionsAreReflectedInMetadata(): void
...
+        );
+
+        $this->connection->executeStatement($createSqls[0]);
+
+        $onlineTable = $this->schemaManager->introspectTable('table_with_index_length');
+
+        $comparator = $this->schemaManager->createComparator();
+        $tablesDiff = $comparator->compareTables($onlineTable, $metadataTable);
+
+        self::assertSame([], $this->connection->getDatabasePlatform()->getAlterTableSQL($tablesDiff));
```
Comment (greg0ire) [inline tests/Functional/Schema/MySQLSchemaManagerTest.php:565]:
```suggestion
        self::assertTrue($tablesDiff->isEmpty());
```

**Comment by derrabus** [comment]
Thank you for this reproducer.

> If any of indexed columns has length equal to the varchar limit, diff always generates index re-creation SQLs (drop + create).

This happens because MySQL (and MariaDB as well, FTR) silently changes the index to the full length of the column. You can reproduce this in plain SQL:

```MySQL
DROP TABLE IF EXISTS foo;
CREATE TABLE foo (col1 VARCHAR(20) NOT NULL, KEY idx_col1 (col1(5)));
SHOW CREATE TABLE foo;
```

The dumped SQL should contain the index definition as we specified it: `KEY idx_col1 (col1(5))`. Now do the same with matching the column length:

```MySQL
DROP TABLE IF EXISTS foo;
CREATE TABLE foo (col1 VARCHAR(20) NOT NULL, KEY idx_col1 (col1(20)));
SHOW CREATE TABLE foo;
```

The returned index definition now reads as `KEY idx_col1 (col1)`.

How would a fix on DBAL's side look like? Should we anticipate this behavior already in the comparator?

> This is easily fixable by passing null for those columns where I want to keep full length.

That would be my recommended workaround, tbh. If you know you're dealing with a MySQL/MariaDB server, then setting the index length to the VARCHAR column length is a futile attempt.

**Comment by janedbal** [comment]
> How would a fix on DBAL's side look like? Should we anticipate this behavior already in the comparator?

I assume yes, since the comparator now has access to the platform used, it should be possible.

**Comment by derrabus** [comment]
Okay. Do you want to give it a try? Otherwise I'd convert this PR into an issue.

**Comment by janedbal** [comment]
I have more serious dbal bugs on my list I want to try fixing once I find some time. 

Since this one has simple workaround, I dont think I will invest into this soon. Feel free to convert it to issue.

**Comment by derrabus** [comment]
Closing in favor of #5989.

**Pull request closed without merging**